File: leetcode/tree/easy/java_solutions/BalancedBinaryTree_110.java
package leetcode.tree.easy.java_solutions;

import static leetcode.tree.Helper.*;

public class BalancedBinaryTree_110 {

    private static class Pair {
        private final int height;
        private final boolean isValid;

        private Pair(int height, boolean isValid) {
            this.height = height;
            this.isValid = isValid;
        }
    }

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        return maxHeight(root).isValid;
    }

    private Pair maxHeight(TreeNode root) {
        if (root == null) {
            return new Pair(0, true);
        }
        Pair left = maxHeight(root.left);
        Pair right = maxHeight(root.right);
        if (Math.abs(left.height - right.height) > 1 || !left.isValid || !right.isValid) {
            return new Pair(0, false);
        }
        return new Pair(Math.max(left.height, right.height) + 1, true);
    }
}

File: leetcode/tree/easy/golang_solutions/balanced_binary_tree_110.go
package golang_solutions

import (
	"github.com/vielendanke/algorithms_in_place/leetcode/tree"
	"math"
)

type pair struct {
	height     int
	isBalanced bool
}

func isBalanced(root *tree.TreeNode) bool {
	if root == nil {
		return true
	}
	return checkIfTreeIsBalanced(root).isBalanced
}

func checkIfTreeIsBalanced(root *tree.TreeNode) *pair {
	if root == nil {
		return &pair{1, true}
	}
	left := checkIfTreeIsBalanced(root.Left)
	right := checkIfTreeIsBalanced(root.Right)

	if math.Abs(float64(left.height-right.height)) > 1 || !left.isBalanced || !right.isBalanced {
		left.isBalanced = false
		return left
	}
	right.height = int(math.Max(float64(left.height), float64(right.height))) + 1
	right.isBalanced = true
	return right
}
